src/evaluators/SwitchStatement.js

```diff
@@ -34,18 +37,217 @@ function CaseSelectorEvaluation(
+      let c = cases[caseIndex];
+      for (let i = 0; i < c.consequent.length && !(result instanceof Completion); ++i) {
+        let node = c.consequent[i];
+        let r = env.evaluateCompletion(node, strictCode);
+
+        // TODO understand Reference and how to deal with it better
+        // TODO come up with a strategy for handling PossiblyNormalCompletion correctly
+        if (r instanceof Reference) {
+          throw ReportNoReferenceError();
+        } else if (r instanceof PossiblyNormalCompletion) {
+          throw ReportNoPossiblyNormalCompletionError();
+        }
+
+        result = GetValue(r, result, realm);
+      }
+      caseIndex++;
+    }
+
+    // TODO work out what to do with Reference
+    if (result instanceof Reference) {
+      throw ReportNoReferenceError();
+    }
+
+    if (result instanceof BreakCompletion) {
+      return result.value;
+    } else if (result instanceof AbruptCompletion) {
+      throw ReportNoAbruptCompletionError();
+    } else if (result instanceof Completion) {
+      return result.value;
+    } else {
+      return result;
+    }
+  };
+
+  let AbstractCaseEvaluation = (caseIndex: number, condition: AbstractValue): Effects => {
+    return Path.withCondition(condition, () => {
+      return realm.evaluateForEffects(() => {
+        return DefiniteCaseEvaluation(caseIndex);
+      });
+    });
+  };
+
+  let ConstructPathCondition = (
+    previousConditions: Array<AbstractValue>,
+    currentCondition: AbstractValue
+  ): AbstractValue => {
+    let fullCondition = currentCondition;
+    for (let previousCondition of previousConditions) {
+      fullCondition = AbstractValue.createFromLogicalOp(realm, "&&", fullCondition, previousCondition);
+    }
+
+    invariant(fullCondition instanceof AbstractValue);
+    return fullCondition;
+  };
+
+  // we remember the inverse of the condition for each path that is considered
+  // these are used to refine the condition for the next path considered as well
+  // as the condition under which the default path should be explored
+  // an abstract view of all the possiblities
+  let pathConditions: Array<AbstractValue> = [];
+  // we join all the values and effects of paths that might be traversed so we get
+  // an abstract view of all the possiblities
+  let collectedEffects: Array<[Effects, AbstractValue]> = [];
+
+  // iterate over each case block, beginning with the first.
+  // evaluate the case selector and see if we can definitively take the case
+  // if so, we are done.
+  // if not, die.
+  for (let i = 0; i < cases.length; ++i) {
+    // only consider the default case once all other cases have been examined
+    if (i === defaultCaseIndex) continue;
+
+    let c = cases[i];
+    let test = c.test;
+    invariant(test);
+
+    let selector = CaseSelectorEvaluation(test, strictCode, env, realm);
+    let selectionResult = computeBinary(realm, "===", input, selector);
```

Comment:
The comment for computeBinary points out: "Note that calling this can result in user code running, which can side-effect the heap.".

This is highly problematic in this context because we may get here AFTER the case that is actually executed at runtime has been processed. In fact, only the FIRST case condition can be evaluated unconditionally.

Consequently, the only safe way to evaluate a list case clauses is in the form of if else if else if ..... This is best expressed in the form of recursive function, not an iteration, since you need a stack of conditions.

Reply:
Ok. I'll kill this PR and try again!